### PR TITLE
EXPERIMENT: port to rowan#34

### PR DESCRIPTION
**Pull request by CAD97**
<details><summary>With this branch</summary>

```powershell
PS D:\usr\Documents\Code\Rust\rust-analyzer> cargo run --bin ra_cli --release -- analysis-stats .
    Finished release [optimized + debuginfo] target(s) in 0.46s
     Running `target\release\ra_cli.exe analysis-stats .`
Database loaded, 221 roots, 1.0601376s
Crates in this dir: 27
Total modules found: 331
Total declarations: 11135
Total functions: 3839
Item Collection: 11.8499283s, 0b allocated 0b resident
Total expressions: 89244
Expressions of unknown type: 6960 (7%)
Expressions of partially unknown type: 3522 (3%)
Type mismatches: 3568
Inference: 36.3460289s, 0b allocated 0b resident
Total: 48.1964408s, 0b allocated 0b resident
PS D:\usr\Documents\Code\Rust\rust-analyzer> Measure-Command { type "D:\rust-lang\src\libcore\unicode\tables.rs" | cargo run --bin ra_cli --release -- symbols }
    Finished release [optimized + debuginfo] target(s) in 0.45s


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 738
Ticks             : 7380228
TotalDays         : 8.54193055555555E-06
TotalHours        : 0.000205006333333333
TotalMinutes      : 0.01230038
TotalSeconds      : 0.7380228
TotalMilliseconds : 738.0228



PS D:\usr\Documents\Code\Rust\rust-analyzer> Measure-Command { type "D:\rust-lang\src\libcore\unicode\tables.rs" | cargo run --bin ra_cli --release -- parse --no-dump } 
    Finished release [optimized + debuginfo] target(s) in 0.44s
     Running `target\release\ra_cli.exe parse --no-dump`


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 603
Ticks             : 6035426
TotalDays         : 6.98544675925926E-06
TotalHours        : 0.000167650722222222
TotalMinutes      : 0.0100590433333333
TotalSeconds      : 0.6035426
TotalMilliseconds : 603.5426



```
</details>

<details><summary>Without this branch (<a href="https://github.com/rust-analyzer/rust-analyzer/commit/5451bfb9">5451bfb9</a>)</summary>

```powershell
PS D:\usr\Documents\Code\Rust\rust-analyzer> cargo run --bin ra_cli --release -- analysis-stats .
    Finished release [optimized + debuginfo] target(s) in 0.45s
     Running `target\release\ra_cli.exe analysis-stats .`
Database loaded, 220 roots, 1.0174838s
Crates in this dir: 27
Total modules found: 331
Total declarations: 11135
Total functions: 3839
Item Collection: 10.509602s, 0b allocated 0b resident
Total expressions: 89241                                                                                                                                                 
Expressions of unknown type: 6959 (7%)
Expressions of partially unknown type: 3522 (3%)
Type mismatches: 3569
Inference: 34.963529s, 0b allocated 0b resident
Total: 45.4737377s, 0b allocated 0b resident
PS D:\usr\Documents\Code\Rust\rust-analyzer> Measure-Command { type "D:\rust-lang\src\libcore\unicode\tables.rs" | cargo run --bin ra_cli --release -- symbols }         
    Finished release [optimized + debuginfo] target(s) in 0.44s


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 587
Ticks             : 5875475
TotalDays         : 6.80031828703704E-06
TotalHours        : 0.000163207638888889
TotalMinutes      : 0.00979245833333333
TotalSeconds      : 0.5875475
TotalMilliseconds : 587.5475



PS D:\usr\Documents\Code\Rust\rust-analyzer> Measure-Command { type "D:\rust-lang\src\libcore\unicode\tables.rs" | cargo run --bin ra_cli --release -- parse --no-dump } 
    Finished release [optimized + debuginfo] target(s) in 0.44s
     Running `target\release\ra_cli.exe parse --no-dump`


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 573
Ticks             : 5737453
TotalDays         : 6.64057060185185E-06
TotalHours        : 0.000159373694444444
TotalMinutes      : 0.00956242166666667
TotalSeconds      : 0.5737453
TotalMilliseconds : 573.7453



```
</details>

I'm not able to do test `--features jemalloc`/`--memory-usage` on Windows, so if someone could do that for me, I'd be appreciative.

Based on the unscientific test runs above, this seems like a net loss. However, this is not necessarily the end state of this avenue of `rowan` size optimization.

### CODE REVIEW

**Comment by matklad** [comment]
In general, I think we should optimize for memory consumption first, and
for speed second. I wonder if it makes sense to do some
back-of-the-envelope calculations of how large, in bytes, the tree will be
with full optimizations applied. This needs some simulating, as interning
of tokens is a crucial bit here.

**Comment by lnicola** [comment]
I'll give it a try once I figure out how to build RA with `jemalloc`.

**Comment by matklad** [comment]
`cargo xtask install --jemalloc` or uncommenting `# features` in
cli/Cargo.toml

**Comment by lnicola** [comment]
> uncommenting `# features` in cli/Cargo.toml

```diff
diff --git i/crates/ra_cli/Cargo.toml w/crates/ra_cli/Cargo.toml
index fcd102e8b..947ca19ea 100644
--- i/crates/ra_cli/Cargo.toml
+++ w/crates/ra_cli/Cargo.toml
@@ -19,4 +19,7 @@ ra_db = { path = "../ra_db" }
 [dependencies.ra_prof]
 path = "../ra_prof"
 # features = [ "cpu_profiler" ]
-# features = [ "jemalloc" ]
+features = [ "jemalloc" ]
+
+[features]
+jemalloc = []
```

`cargo run --features jemalloc --release -- analysis-stats --memory-usage ../..`

Ugh.

**Comment by lnicola** [comment]
Basically, it's.. too noisy:

<details><summary>`boxed-tokens`</summary>
Database loaded, 221 roots, 234.860189ms
Crates in this dir: 27
Total modules found: 331
Total declarations: 11135
Total functions: 3839
Item Collection: 7.921741849s, 356mb allocated 402mb resident
Total expressions: 89244
Expressions of unknown type: 6962 (7%)
Expressions of partially unknown type: 3523 (3%)
Type mismatches: 3571
Inference: 15.611964896s, 1017mb allocated 1063mb resident
Total: 23.533924845s, 1017mb allocated 1063mb resident
   154mb CrateDefMapQuery
   109mb TraitSolveQuery (deps)
    72mb ImplDatumQuery (deps)
    42mb InferQuery (deps)
    40mb MacroArgQuery
    32mb ImplDatumQuery
    26mb RawItemsQuery
    22mb ParseQuery
    20mb BodyHirQuery
    19mb InferQuery
    12mb ImplsInModuleQuery
    11mb GenericParamsQuery (deps)
     8mb ImplsForTraitQuery (deps)
     8mb GenericParamsQuery
     7mb ImplsForTraitQuery
     7mb BodyWithSourceMapQuery
     6mb MacroArgQuery (deps)
     6mb MacroExpandQuery (deps)
     6mb ParseMacroQuery (deps)
     6mb AstIdMapQuery (deps)
     5mb RawItemsQuery (deps)
     5mb ImplsInModuleWithSourceMapQuery (deps)
     5mb RawItemsWithSourceMapQuery (deps)
     5mb MacroDefQuery
     4mb AstIdMapQuery
     4mb GenericPredicatesQuery (deps)
     4mb TypeForDefQuery (deps)
     4mb CrateDefMapQuery (deps)
  2933kb ExprScopesQuery
  2884kb BodyWithSourceMapQuery (deps)
  2550kb GenericDefaultsQuery (deps)
  2531kb FnDataQuery
  2504kb ImplsInCrateQuery (deps)
  2258kb CallableItemSignatureQuery (deps)
  1786kb BodyHirQuery (deps)
  1641kb FnDataQuery (deps)
  1429kb RawItemsWithSourceMapQuery
  1307kb TypeForDefQuery
  1301kb ExprScopesQuery (deps)
  1044kb ImplsInModuleQuery (deps)
  1026kb GenericPredicatesQuery
   952kb CallableItemSignatureQuery
   894kb ImplsInModuleWithSourceMapQuery
   837kb TypeForFieldQuery (deps)
   834kb ParseQuery (deps)
   711kb StructDatumQuery (deps)
   642kb TypeAliasDataQuery (deps)
   546kb StructDatumQuery
   519kb TypeAliasDataQuery
   434kb ImplsInCrateQuery
   303kb ModuleLangItemsQuery (deps)
   289kb GenericDefaultsQuery
   275kb EnumDataQuery
   257kb StructDataQuery
   244kb TraitSolveQuery
   238kb TraitDatumQuery
   196kb TraitDatumQuery (deps)
   188kb MacroDefQuery (deps)
   172kb StructDataQuery (deps)
   148kb TraitItemsIndexQuery (deps)
   132kb ParseMacroQuery
    83kb LangItemQuery (deps)
    82kb EnumDataQuery (deps)
    79kb TraitDataQuery (deps)
    73kb TypeForFieldQuery
    59kb MacroExpandQuery
    27kb TraitDataQuery
    19kb StaticDataQuery (deps)
     8kb ConstDataQuery (deps)
   4096b TraitItemsIndexQuery
   4096b CrateLangItemsQuery
   4096b AssociatedTyDataQuery (deps)
    128b StaticDataQuery
      0b SourceRootCratesQuery
      0b SourceRootCratesQuery (deps)
      0b ConstDataQuery
      0b ModuleLangItemsQuery
      0b CrateLangItemsQuery (deps)
      0b LangItemQuery
      0b DocumentationQuery
      0b DocumentationQuery (deps)
      0b AssociatedTyDataQuery
leftover: 324mb


Database loaded, 221 roots, 232.368974ms
Crates in this dir: 27
Total modules found: 331
Total declarations: 11135
Total functions: 3839
Item Collection: 7.98660706s, 356mb allocated 392mb resident
Total expressions: 89244
Expressions of unknown type: 6960 (7%)
Expressions of partially unknown type: 3521 (3%)
Type mismatches: 3571
Inference: 15.307437211s, 1011mb allocated 1059mb resident
Total: 23.2942586s, 1011mb allocated 1059mb resident
   154mb CrateDefMapQuery
   109mb TraitSolveQuery (deps)
    72mb ImplDatumQuery (deps)
    42mb InferQuery (deps)
    40mb MacroArgQuery
    32mb ImplDatumQuery
    26mb RawItemsQuery
    20mb ParseQuery
    20mb BodyHirQuery
    19mb InferQuery
    12mb ImplsInModuleQuery
    11mb GenericParamsQuery (deps)
     8mb ImplsForTraitQuery (deps)
     7mb GenericParamsQuery
     7mb ImplsForTraitQuery
     7mb BodyWithSourceMapQuery
     6mb MacroArgQuery (deps)
     6mb MacroExpandQuery (deps)
     6mb ParseMacroQuery (deps)
     6mb AstIdMapQuery (deps)
     5mb RawItemsQuery (deps)
     5mb ImplsInModuleWithSourceMapQuery (deps)
     5mb RawItemsWithSourceMapQuery (deps)
     4mb MacroDefQuery
     4mb AstIdMapQuery
     4mb GenericPredicatesQuery (deps)
     4mb CrateDefMapQuery (deps)
     4mb TypeForDefQuery (deps)
  2963kb ExprScopesQuery
  2668kb BodyWithSourceMapQuery (deps)
  2533kb GenericDefaultsQuery (deps)
  2523kb FnDataQuery
  2484kb ImplsInCrateQuery (deps)
  2266kb CallableItemSignatureQuery (deps)
  1794kb BodyHirQuery (deps)
  1649kb FnDataQuery (deps)
  1430kb RawItemsWithSourceMapQuery
  1226kb ExprScopesQuery (deps)
  1213kb TypeForDefQuery
  1048kb ImplsInModuleQuery (deps)
  1025kb GenericPredicatesQuery
   924kb CallableItemSignatureQuery
   868kb TypeForFieldQuery (deps)
   760kb ImplsInModuleWithSourceMapQuery
   738kb ParseQuery (deps)
   642kb TypeAliasDataQuery (deps)
   586kb StructDatumQuery (deps)
   521kb TypeAliasDataQuery
   455kb StructDatumQuery
   316kb ImplsInCrateQuery
   304kb ModuleLangItemsQuery (deps)
   277kb TraitSolveQuery
   275kb EnumDataQuery
   255kb StructDataQuery
   254kb GenericDefaultsQuery
   243kb TraitDatumQuery
   213kb TraitDatumQuery (deps)
   172kb StructDataQuery (deps)
   163kb MacroDefQuery (deps)
   148kb TraitItemsIndexQuery (deps)
    83kb LangItemQuery (deps)
    82kb EnumDataQuery (deps)
    79kb TraitDataQuery (deps)
    77kb ParseMacroQuery
    76kb TypeForFieldQuery
    73kb MacroExpandQuery
    27kb TraitDataQuery
    19kb StaticDataQuery (deps)
     8kb StaticDataQuery
   4096b TraitItemsIndexQuery
   4096b ConstDataQuery (deps)
   4096b CrateLangItemsQuery (deps)
   4096b AssociatedTyDataQuery (deps)
      0b SourceRootCratesQuery
      0b SourceRootCratesQuery (deps)
      0b ConstDataQuery
      0b ModuleLangItemsQuery
      0b CrateLangItemsQuery
      0b LangItemQuery
      0b DocumentationQuery
      0b DocumentationQuery (deps)
      0b AssociatedTyDataQuery
leftover: 320mb
</details>
<details><summary>`boxed-tokens^`</summary>
Database loaded, 220 roots, 240.855792ms
Crates in this dir: 27
Total modules found: 331
Total declarations: 11135
Total functions: 3839
Item Collection: 7.230829516s, 358mb allocated 399mb resident
Total expressions: 89241
Expressions of unknown type: 6962 (7%)
Expressions of partially unknown type: 3522 (3%)
Type mismatches: 3577
Inference: 13.914667895s, 974mb allocated 1018mb resident
Total: 21.145667983s, 974mb allocated 1018mb resident
   154mb CrateDefMapQuery
   110mb TraitSolveQuery (deps)
    72mb ImplDatumQuery (deps)
    42mb InferQuery (deps)
    40mb MacroArgQuery
    32mb ImplDatumQuery
    29mb ParseQuery
    26mb RawItemsQuery
    20mb BodyHirQuery
    18mb InferQuery
    12mb ImplsInModuleQuery
    11mb GenericParamsQuery (deps)
     8mb ImplsForTraitQuery (deps)
     8mb GenericParamsQuery
     7mb ImplsForTraitQuery
     7mb BodyWithSourceMapQuery
     6mb MacroArgQuery (deps)
     6mb AstIdMapQuery (deps)
     6mb MacroExpandQuery (deps)
     6mb ParseMacroQuery (deps)
     5mb RawItemsQuery (deps)
     5mb RawItemsWithSourceMapQuery (deps)
     5mb ImplsInModuleWithSourceMapQuery (deps)
     4mb MacroDefQuery
     4mb AstIdMapQuery
     4mb GenericPredicatesQuery (deps)
     4mb TypeForDefQuery (deps)
     4mb CrateDefMapQuery (deps)
  2932kb ExprScopesQuery
  2802kb BodyWithSourceMapQuery (deps)
  2528kb GenericDefaultsQuery (deps)
  2519kb FnDataQuery
  2497kb ImplsInCrateQuery (deps)
  2255kb CallableItemSignatureQuery (deps)
  1641kb FnDataQuery (deps)
  1485kb BodyHirQuery (deps)
  1435kb RawItemsWithSourceMapQuery
  1281kb ExprScopesQuery (deps)
  1187kb TypeForDefQuery
  1039kb ImplsInModuleQuery (deps)
  1012kb GenericPredicatesQuery
   969kb CallableItemSignatureQuery
   807kb TypeForFieldQuery (deps)
   766kb ImplsInModuleWithSourceMapQuery
   728kb ParseQuery (deps)
   637kb TypeAliasDataQuery (deps)
   573kb StructDatumQuery (deps)
   516kb TypeAliasDataQuery
   453kb StructDatumQuery
   338kb ImplsInCrateQuery
   296kb EnumDataQuery
   295kb GenericDefaultsQuery
   289kb ModuleLangItemsQuery (deps)
   264kb StructDataQuery
   237kb TraitSolveQuery
   226kb TraitDatumQuery
   215kb MacroDefQuery (deps)
   199kb TraitDatumQuery (deps)
   165kb StructDataQuery (deps)
   146kb TraitItemsIndexQuery (deps)
   120kb TypeForFieldQuery
    87kb EnumDataQuery (deps)
    78kb LangItemQuery (deps)
    56kb TraitDataQuery (deps)
    27kb ParseMacroQuery
    19kb TraitDataQuery
    19kb StaticDataQuery (deps)
    19kb AssociatedTyDataQuery (deps)
    15kb ConstDataQuery
     8kb ConstDataQuery (deps)
     4kb StaticDataQuery
   4096b TraitItemsIndexQuery
   4096b ModuleLangItemsQuery
   4096b CrateLangItemsQuery (deps)
      0b SourceRootCratesQuery
      0b SourceRootCratesQuery (deps)
      0b MacroExpandQuery
      0b CrateLangItemsQuery
      0b LangItemQuery
      0b DocumentationQuery
      0b DocumentationQuery (deps)
      0b AssociatedTyDataQuery
leftover: 274mb

Database loaded, 220 roots, 232.062315ms
Crates in this dir: 27
Total modules found: 331
Total declarations: 11135
Total functions: 3839
Item Collection: 7.261006966s, 358mb allocated 399mb resident
Total expressions: 89241
Expressions of unknown type: 6964 (7%)
Expressions of partially unknown type: 3524 (3%)
Type mismatches: 3577
Inference: 14.242552289s, 984mb allocated 1027mb resident
Total: 21.50378053s, 984mb allocated 1027mb resident
   154mb CrateDefMapQuery
   109mb TraitSolveQuery (deps)
    72mb ImplDatumQuery (deps)
    42mb InferQuery (deps)
    40mb MacroArgQuery
    32mb ImplDatumQuery
    30mb ParseQuery
    26mb RawItemsQuery
    20mb BodyHirQuery
    19mb InferQuery
    12mb ImplsInModuleQuery
    11mb GenericParamsQuery (deps)
     8mb ImplsForTraitQuery (deps)
     8mb GenericParamsQuery
     7mb ImplsForTraitQuery
     7mb BodyWithSourceMapQuery
     6mb MacroArgQuery (deps)
     6mb AstIdMapQuery (deps)
     6mb MacroExpandQuery (deps)
     6mb ParseMacroQuery (deps)
     5mb RawItemsQuery (deps)
     5mb ImplsInModuleWithSourceMapQuery (deps)
     5mb RawItemsWithSourceMapQuery (deps)
     4mb MacroDefQuery
     4mb AstIdMapQuery
     4mb GenericPredicatesQuery (deps)
     4mb TypeForDefQuery (deps)
     4mb CrateDefMapQuery (deps)
  2974kb BodyWithSourceMapQuery (deps)
  2909kb ExprScopesQuery
  2566kb GenericDefaultsQuery (deps)
  2537kb FnDataQuery
  2468kb ImplsInCrateQuery (deps)
  2259kb CallableItemSignatureQuery (deps)
  1724kb BodyHirQuery (deps)
  1639kb FnDataQuery (deps)
  1447kb RawItemsWithSourceMapQuery
  1302kb ExprScopesQuery (deps)
  1014kb TypeForDefQuery
   997kb GenericPredicatesQuery
   930kb ImplsInModuleQuery (deps)
   922kb CallableItemSignatureQuery
   871kb TypeForFieldQuery (deps)
   861kb ImplsInModuleWithSourceMapQuery
   749kb ParseQuery (deps)
   648kb StructDatumQuery (deps)
   633kb TypeAliasDataQuery (deps)
   612kb ImplsInCrateQuery
   503kb TypeAliasDataQuery
   427kb StructDatumQuery
   359kb MacroDefQuery (deps)
   320kb ModuleLangItemsQuery (deps)
   287kb EnumDataQuery
   273kb GenericDefaultsQuery
   270kb StructDataQuery
   258kb TraitSolveQuery
   257kb TraitDatumQuery
   202kb TraitDatumQuery (deps)
   153kb StructDataQuery (deps)
   122kb TraitItemsIndexQuery (deps)
    92kb TypeForFieldQuery
    85kb EnumDataQuery (deps)
    79kb TraitDataQuery (deps)
    77kb LangItemQuery (deps)
    63kb ParseMacroQuery
    23kb AssociatedTyDataQuery (deps)
    23kb TraitDataQuery
    15kb StaticDataQuery
    14kb ConstDataQuery
     8kb ConstDataQuery (deps)
   4096b MacroExpandQuery
   4096b TraitItemsIndexQuery
   4096b StaticDataQuery (deps)
   4096b ModuleLangItemsQuery
   4096b CrateLangItemsQuery (deps)
      0b SourceRootCratesQuery
      0b SourceRootCratesQuery (deps)
      0b CrateLangItemsQuery
      0b LangItemQuery
      0b DocumentationQuery
      0b DocumentationQuery (deps)
      0b AssociatedTyDataQuery
leftover: 283mb
</details>

**Comment by lnicola** [comment]
TL;DR:

```
boxed-tokens
Total: 23.533924845s, 1017mb allocated 1063mb resident
Total: 23.2942586s, 1011mb allocated 1059mb resident

boxed-tokens^
Total: 21.145667983s, 974mb allocated 1018mb resident
Total: 21.50378053s, 984mb allocated 1027mb resident
```

**Comment by lnicola** [comment]
On the other hand, `ParseQuery` went from 29-30 MB to 20-22 MB. Not sure how relevant it is.

And for some reason, there's an extra root with the new code.

**Comment by lnicola** [comment]
I re-ran the test using the same RA version:

<details>

```
boxed-tokens
Database loaded, 220 roots, 242.310731ms
Crates in this dir: 27
Total modules found: 332
Total declarations: 11174
Total functions: 3859
Item Collection: 7.930623453s, 356mb allocated 389mb resident
Total expressions: 89607
Expressions of unknown type: 6964 (7%)
Expressions of partially unknown type: 3533 (3%)
Type mismatches: 3608
Inference: 14.96352302s, 964mb allocated 1008mb resident
Total: 22.894338156s, 964mb allocated 1008mb resident
   153mb CrateDefMapQuery
   110mb TraitSolveQuery (deps)
    72mb ImplDatumQuery (deps)
    42mb InferQuery (deps)
    40mb MacroArgQuery
    32mb ImplDatumQuery
    26mb RawItemsQuery
    20mb BodyHirQuery
    19mb InferQuery
    18mb ParseQuery
    12mb ImplsInModuleQuery
    11mb GenericParamsQuery (deps)
     8mb ImplsForTraitQuery (deps)
     8mb GenericParamsQuery
     7mb ImplsForTraitQuery
     7mb BodyWithSourceMapQuery
     6mb MacroArgQuery (deps)
     6mb MacroExpandQuery (deps)
     6mb ParseMacroQuery (deps)
     6mb AstIdMapQuery (deps)
     5mb RawItemsQuery (deps)
     5mb ImplsInModuleWithSourceMapQuery (deps)
     5mb RawItemsWithSourceMapQuery (deps)
     4mb MacroDefQuery
     4mb GenericPredicatesQuery (deps)
     4mb AstIdMapQuery
     4mb CrateDefMapQuery (deps)
     4mb TypeForDefQuery (deps)
  2967kb ExprScopesQuery
  2754kb BodyWithSourceMapQuery (deps)
  2546kb ImplsInCrateQuery (deps)
  2540kb FnDataQuery
  2512kb GenericDefaultsQuery (deps)
  2274kb CallableItemSignatureQuery (deps)
  1964kb BodyHirQuery (deps)
  1663kb FnDataQuery (deps)
  1434kb RawItemsWithSourceMapQuery
  1394kb TypeForDefQuery
  1232kb ExprScopesQuery (deps)
  1062kb ImplsInModuleQuery (deps)
  1008kb GenericPredicatesQuery
   945kb ImplsInModuleWithSourceMapQuery
   932kb CallableItemSignatureQuery
   854kb TypeForFieldQuery (deps)
   705kb ParseQuery (deps)
   632kb TypeAliasDataQuery (deps)
   592kb StructDatumQuery (deps)
   496kb TypeAliasDataQuery
   456kb StructDatumQuery
   297kb ModuleLangItemsQuery (deps)
   278kb ImplsInCrateQuery
   275kb EnumDataQuery
   269kb GenericDefaultsQuery
   260kb StructDataQuery
   245kb TraitSolveQuery
   226kb TraitDatumQuery
   223kb ParseMacroQuery
   202kb TraitDatumQuery (deps)
   188kb MacroDefQuery (deps)
   165kb StructDataQuery (deps)
   151kb MacroExpandQuery
   134kb TraitItemsIndexQuery (deps)
   107kb TypeForFieldQuery
    90kb TraitDataQuery (deps)
    82kb EnumDataQuery (deps)
    78kb LangItemQuery (deps)
    27kb TraitDataQuery
    23kb ConstDataQuery (deps)
    23kb AssociatedTyDataQuery (deps)
    19kb StaticDataQuery (deps)
   4096b TraitItemsIndexQuery
   4096b ModuleLangItemsQuery
   4096b CrateLangItemsQuery (deps)
   3840b ConstDataQuery
      0b SourceRootCratesQuery
      0b SourceRootCratesQuery (deps)
      0b StaticDataQuery
      0b CrateLangItemsQuery
      0b LangItemQuery
      0b DocumentationQuery
      0b DocumentationQuery (deps)
      0b AssociatedTyDataQuery
leftover: 274mb

boxed-tokens^
Database loaded, 220 roots, 244.833197ms
Crates in this dir: 27
Total modules found: 332
Total declarations: 11174
Total functions: 3859
Item Collection: 7.229507785s, 363mb allocated 407mb resident
Total expressions: 89607
Expressions of unknown type: 6962 (7%)
Expressions of partially unknown type: 3531 (3%)
Type mismatches: 3608
Inference: 14.382722918s, 1022mb allocated 1069mb resident
Total: 21.612443837s, 1022mb allocated 1069mb resident
   153mb CrateDefMapQuery
   109mb TraitSolveQuery (deps)
    72mb ImplDatumQuery (deps)
    42mb InferQuery (deps)
    40mb MacroArgQuery
    32mb ImplDatumQuery
    30mb ParseQuery
    26mb RawItemsQuery
    20mb BodyHirQuery
    19mb InferQuery
    13mb ImplsInModuleQuery
    11mb GenericParamsQuery (deps)
     8mb ImplsForTraitQuery (deps)
     8mb GenericParamsQuery
     7mb ImplsForTraitQuery
     7mb BodyWithSourceMapQuery
     6mb MacroArgQuery (deps)
     6mb MacroExpandQuery (deps)
     6mb AstIdMapQuery (deps)
     6mb ParseMacroQuery (deps)
     5mb RawItemsQuery (deps)
     5mb ImplsInModuleWithSourceMapQuery (deps)
     5mb RawItemsWithSourceMapQuery (deps)
     5mb MacroDefQuery
     4mb TypeForDefQuery (deps)
     4mb AstIdMapQuery
     4mb GenericPredicatesQuery (deps)
     4mb CrateDefMapQuery (deps)
  2980kb ExprScopesQuery
  2959kb BodyWithSourceMapQuery (deps)
  2549kb FnDataQuery
  2531kb GenericDefaultsQuery (deps)
  2467kb ImplsInCrateQuery (deps)
  2259kb CallableItemSignatureQuery (deps)
  1635kb FnDataQuery (deps)
  1623kb BodyHirQuery (deps)
  1427kb RawItemsWithSourceMapQuery
  1281kb ExprScopesQuery (deps)
  1018kb GenericPredicatesQuery
  1000kb ImplsInModuleQuery (deps)
   947kb TypeForDefQuery
   933kb CallableItemSignatureQuery
   840kb TypeForFieldQuery (deps)
   806kb ParseQuery (deps)
   790kb ImplsInModuleWithSourceMapQuery
   657kb TypeAliasDataQuery (deps)
   655kb StructDatumQuery (deps)
   512kb TypeAliasDataQuery
   451kb ImplsInCrateQuery
   418kb StructDatumQuery
   302kb EnumDataQuery
   298kb GenericDefaultsQuery
   289kb ModuleLangItemsQuery (deps)
   269kb StructDataQuery
   262kb TraitSolveQuery
   243kb TraitDatumQuery
   216kb TraitDatumQuery (deps)
   189kb MacroDefQuery (deps)
   154kb ParseMacroQuery
   150kb StructDataQuery (deps)
   146kb TraitItemsIndexQuery (deps)
    78kb LangItemQuery (deps)
    78kb TypeForFieldQuery
    74kb TraitDataQuery (deps)
    72kb EnumDataQuery (deps)
    28kb MacroExpandQuery
    27kb TraitDataQuery
    19kb StaticDataQuery (deps)
    19kb AssociatedTyDataQuery (deps)
     8kb ConstDataQuery (deps)
   4096b TraitItemsIndexQuery
   4096b CrateLangItemsQuery (deps)
      0b SourceRootCratesQuery
      0b SourceRootCratesQuery (deps)
      0b ConstDataQuery
      0b StaticDataQuery
      0b ModuleLangItemsQuery
      0b CrateLangItemsQuery
      0b LangItemQuery
      0b DocumentationQuery
      0b DocumentationQuery (deps)
      0b AssociatedTyDataQuery
leftover: 320mb
```

</details>

```
boxed-tokens
Total: 23.64864768s, 1027mb allocated 1073mb resident
Total: 22.894338156s, 964mb allocated 1008mb resident
Total: 22.958151875s, 968mb allocated 1013mb resident

boxed-tokens^
Total: 22.062337695s, 1029mb allocated 1076mb resident
Total: 21.612443837s, 1022mb allocated 1069mb resident
Total: 21.77718634s, 1030mb allocated 1077mb resident
```

**Comment by CAD97** [comment]
Superseded by #2182.
